File: sentimental-credit/credit.py
import re
from math import floor

input = input("Number: ")
# input = "378282246310005"
card = int(re.findall(r"\d+", input)[0])

# number list that will be multiplied
firsts_sum = 0

# sum of the second lit of numbers only for last addition
seconds = []

# position and length tracker
pos = 0

# get last digit and append to the list or sum
OGcard = card
while card > 0:
    digit = card % 10

    if pos % 2 == 0:
        firsts_sum += digit
    else:
        seconds.append(digit)

    card = floor(card / 10)
    pos += 1

card = OGcard

# multiply the firsts list with 2 ensuring that any 2dgt is separated
seconds_sum = 0
for i in range(len(seconds)):
    x = seconds[i]
    x *= 2
    if x > 9:
        last_digit = x % 10
        last_digit += floor(x / 10)
        seconds_sum += last_digit
    else:
        seconds_sum += x

checksum = seconds_sum + firsts_sum
cs_last = checksum % 10

# retrieve card service param digits for validation
first_one = card
first_two = card
while first_one >= 10:
    first_one /= 10

while first_two >= 100:
    first_two /= 10

# floor the nums to remove decimal
first_one = floor(first_one)
first_two = floor(first_two)

# identify card service
clen = pos

if cs_last == 0:
    if clen == 15 and first_two in [34, 37]:
        print("AMEX")
    elif clen == 16 and first_two in range(51, 55+1):
        print("MASTERCARD")
    elif clen in [13, 16] and first_one == 4:
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // long OGnum = get_long("Number: ");
    // long num = OGnum;

    long num = 4012888888881881;
    long OGnum = num;
    int firsts[8] = {};
    int seconds[8] = {};
    int index1 = 0;
    int index2 = 0;
    int pos = 0;

    while (num > 0)
    {
        int digit = num % 10;

        if (pos % 2 == 0)
        {
            firsts[index1++] = digit;
        }
        else
        {
            seconds[index2++] = digit;
        }

        num /= 10;
        pos++;
    }

    int Sum = 0;

    for (int i = 0; i < index2; i++)
    {
        int dbl = seconds[i] * 2;
        if (dbl >= 10)
        {
            Sum += (dbl % 10) + (dbl / 10); // sum the digits of the 2-digit number
        }
        else
        {
            Sum += dbl;
        }
    }
    for (int i = 0; i < index1; i++)
    {
        Sum += firsts[i];
    }

    int sumLastDig = Sum % 10;

    // Checker For Card Service
    long FirstOne = OGnum;
    long FirstTwo = OGnum;
    // printf("%ld/n",OGnum);

    while (FirstOne >= 10)
    {
        FirstOne /= 10;
    }
    while (FirstTwo >= 100)
    {
        FirstTwo /= 10;
    }

    int cardLength = index1 + index2;

    if (sumLastDig == 0 && cardLength == 15 && (FirstTwo == 34 || FirstTwo == 37))
    {
        printf("AMEX\n");
    }
    else if (sumLastDig == 0 && cardLength == 16 && FirstTwo >= 51 && FirstTwo <= 55)
    {
        printf("MASTERCARD\n");
    }
    else if (sumLastDig == 0 && FirstOne == 4 && (cardLength == 13 || cardLength == 16))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}
